File: src/main/java/com/fireblazerrr/survivorbot/chatter/StandardChatter.java
package com.fireblazerrr.survivorbot.chatter;

import com.fireblazerrr.survivorbot.SurvivorBot;
import com.fireblazerrr.survivorbot.channel.Channel;
import com.fireblazerrr.survivorbot.utils.message.MessageNotFoundException;
import com.fireblazerrr.survivorbot.utils.message.Messaging;
import org.bukkit.ChatColor;
import org.bukkit.entity.Player;

import java.util.HashSet;
import java.util.Set;

public class StandardChatter implements Chatter {
    private final Player player;
    private Chatter lastPMSource;
    private Channel activeChannel;
    private Channel lastActiveChannel;
    private Channel lastFocusableChannel;
    private ChatterStorage storage;
    private Set<Channel> channels = new HashSet<>();
    private Set<String> ignores = new HashSet<>();
    private String afkMessage = "";
    private boolean muted = false;
    private boolean afk = false;

    StandardChatter(ChatterStorage storage, Player player) {
        this.storage = storage;
        this.player = player;
    }

    public boolean addChannel(Channel channel, boolean announce, boolean flagUpdate) {
        if (this.channels.contains(channel)) {
            return false;
        } else {
            this.channels.add(channel);
            if (!channel.isMember(this)) {
                channel.addMember(this, announce, flagUpdate);
            }

            if (flagUpdate) {
                this.storage.flagUpdate(this);
            }

            return true;
        }
    }

    public void attachStorage(ChatterStorage storage) {
        this.storage = storage;
    }

    public Result canBan(Channel channel) {
        return SurvivorBot.hasChannelPermission(this.player, channel, Permission.BAN) ?
                Result.ALLOWED :
                (channel.isModerator(this.player.getName()) && SurvivorBot.getChannelManager()
                        .checkModPermission(Permission.BAN) ?
                        Result.ALLOWED :
                        Result.NO_PERMISSION);
    }

    @Override
    public Result canColorMessage(Channel channel, ChatColor color) {
        return null;
    }

    public Result canColorMessages(Channel channel, ChatColor color) {
        return SurvivorBot.hasChannelPermission(this.player, channel, Permission.COLOR) ? Result.ALLOWED :
                (!SurvivorBot.hasChannelPermission(this.player, channel, Permission.valueOf(color.name())) ?
                        Result.NO_PERMISSION : Result.ALLOWED);
    }

    public Result canEmote(Channel channel) {
        return channel == null ? Result.NO_CHANNEL : (!channel.isMember(this) ? Result.INVALID :
                (!channel.isTransient() && SurvivorBot.hasChannelPermission(this.player, channel, Permission.EMOTE) ?
                        (!this.muted && !channel.isMuted(this.player.getName()) ?
                                (!channel.hasWorld(this.player.getWorld()) ? Result.BAD_WORLD : Result.ALLOWED) :
                                Result.MUTED) : Result.NO_PERMISSION));
    }

    public Result canFocus(Channel channel) {
        if (!SurvivorBot.hasChannelPermission(this.player, channel, Permission.FOCUS)) {
            return Result.NO_PERMISSION;
        } else {
            Result speak = this.canSpeak(channel);
            return speak != Result.ALLOWED && speak != Result.INVALID ? Result.NO_PERMISSION : Result.ALLOWED;
        }
    }

    public Result canJoin(Channel channel, String password) {
        return channel.isMember(this) ? Result.INVALID :
                (!SurvivorBot.hasChannelPermission(this.player, channel, Permission.JOIN) ? Result.NO_PERMISSION :
                        (channel.isBanned(this.player.getName()) ? Result.BANNED :
                                (!password.equals(channel.getPassword()) ? Result.BAD_PASSWORD : Result.ALLOWED)));
    }

    public Result canKick(Channel channel) {
        return SurvivorBot.hasChannelPermission(this.player, channel, Permission.KICK) ? Result.ALLOWED :
                (channel.isModerator(this.player.getName()) && SurvivorBot.getChannelManager()
                        .checkModPermission(Permission.KICK) ? Result.ALLOWED : Result.NO_PERMISSION);
    }

    public Result canLeave(Channel channel) {
        return !channel.isMember(this) ? Result.INVALID :
                (!SurvivorBot.hasChannelPermission(this.player, channel, Permission.LEAVE) ? Result.NO_PERMISSION :
                        Result.ALLOWED);
    }

    public Result canModify(String setting, Channel channel) {
        setting = setting.toLowerCase();
        Permission permission;
        switch (setting) {
            case "nick":
                permission = Permission.MODIFY_NICK;
                break;
            case "format":
                permission = Permission.MODIFY_FORMAT;
                break;
            case "distance":
                permission = Permission.MODIFY_DISTANCE;
                break;
            case "color":
                permission = Permission.MODIFY_COLOR;
                break;
            case "shortcut":
                permission = Permission.MODIFY_SHORTCUT;
                break;
            case "password":
                permission = Permission.MODIFY_PASSWORD;
                break;
            case "verbose":
                permission = Permission.MODIFY_VERBOSE;
                break;
            case "chatcost":
                permission = Permission.MODIFY_CHATCOST;
                break;
            case "crossworld":
                permission = Permission.MODIFY_CROSSWORLD;
                break;
            default:
                if (!setting.equals("focusable")) {
                    return Result.INVALID;
                }

                permission = Permission.MODIFY_FOCUSABLE;
                break;
        }

        return SurvivorBot.hasChannelPermission(this.player, channel, permission) ? Result.ALLOWED :
                (channel.isModerator(this.player.getName()) && SurvivorBot.getChannelManager()
                        .checkModPermission(permission) ? Result.ALLOWED : Result.NO_PERMISSION);
    }

    public Result canMute(Channel channel) {
        return SurvivorBot.hasChannelPermission(this.player, channel, Permission.MUTE) ? Result.ALLOWED :
                (channel.isModerator(this.player.getName()) && SurvivorBot.getChannelManager()
                        .checkModPermission(Permission.BAN) ? Result.ALLOWED : Result.NO_PERMISSION);
    }

    public Result canRemove(Channel channel) {
        return SurvivorBot.hasChannelPermission(this.player, channel, Permission.REMOVE) ? Result.ALLOWED :
                (channel.isModerator(this.player.getName()) && SurvivorBot.getChannelManager()
                        .checkModPermission(Permission.REMOVE) ? Result.ALLOWED : Result.NO_PERMISSION);
    }

    public Result canSpeak(Channel channel) {
        return !channel.isMember(this) ? Result.INVALID :
                (!channel.isTransient() && !SurvivorBot.hasChannelPermission(this.player, channel, Permission.SPEAK) ?
                        Result.NO_PERMISSION : (!this.muted && !channel.isMuted(this.player.getName()) ?
                        (!channel.hasWorld(this.player.getWorld()) ? Result.BAD_WORLD : Result.ALLOWED) :
                        Result.MUTED));
    }

    public Result canViewInfo(Channel channel) {
        return !SurvivorBot.hasChannelPermission(this.player, channel, Permission.INFO) ? Result.NO_PERMISSION :
                Result.ALLOWED;
    }

    public boolean equals(Object other) {
        return other == this || (other != null && (other instanceof Chatter && this.player
                .equals(((Chatter) other).getPlayer())));
    }

    public Channel getActiveChannel() {
        return this.activeChannel;
    }

    public String getAFKMessage() {
        return this.afkMessage;
    }

    public Set<Channel> getChannels() {
        return this.channels;
    }

    public Set<String> getIgnores() {
        return this.ignores;
    }

    public Channel getLastActiveChannel() {
        return this.lastActiveChannel;
    }

    public Channel getLastFocusableChannel() {
        return this.lastFocusableChannel;
    }

    public Chatter getLastPrivateMessageSource() {
        return this.lastPMSource;
    }

    public String getName() {
        return this.player.getName();
    }

    public Player getPlayer() {
        return this.player;
    }

    public ChatterStorage getStorage() {
        return this.storage;
    }

    public boolean hasChannel(Channel channel) {
        return this.channels.contains(channel);
    }

    public int hashCode() {
        return this.player.hashCode();
    }

    public boolean isAFK() {
        return this.afk;
    }

    public boolean isIgnoring(Chatter other) {
        return this.isIgnoring(other.getName());
    }

    public boolean isIgnoring(String name) {
        return this.ignores.contains(name.toLowerCase());
    }

    public boolean isInRange(Chatter other, int distance) {
        Player otherPlayer = other.getPlayer();
        return this.player.getWorld().equals(otherPlayer.getWorld()) && this.player.getLocation()
                .distanceSquared(otherPlayer.getLocation()) <= (double) (distance * distance);
    }

    public boolean isMuted() {
        return this.muted;
    }

    public boolean removeChannel(Channel channel, boolean announce, boolean flagUpdate) {
        if (!this.channels.contains(channel)) {
            return false;
        } else {
            this.channels.remove(channel);
            if (channel.isMember(this)) {
                channel.removeMember(this, announce, flagUpdate);
            }

            if (flagUpdate) {
                this.storage.flagUpdate(this);
            }

            return true;
        }
    }

    public void setActiveChannel(Channel channel, boolean announce, boolean flagUpdate) {
        if (channel == null || !channel.equals(this.activeChannel)) {
            if (this.activeChannel != null) {
                this.activeChannel.onFocusLoss(this);
            }

            if (this.activeChannel != null && !this.activeChannel.isTransient()) {
                this.lastActiveChannel = this.activeChannel;
                if (this.canFocus(this.activeChannel) == Result.ALLOWED) {
                    this.lastFocusableChannel = this.activeChannel;
                }
            }

            this.activeChannel = channel;
            if (this.activeChannel != null) {
                this.activeChannel.onFocusGain(this);
                if (announce) {
                    try {
                        Messaging.send(this.player, SurvivorBot.getMessage("chatter_focus"),
                                channel.getColor() + channel.getName());
                    } catch (MessageNotFoundException e) {
                        SurvivorBot.severe("Messages.properties is missing: chatter_focus");
                    }
                }
            }

            if (flagUpdate) {
                this.storage.flagUpdate(this);
            }
        }
    }

    public void setAFK(boolean afk) {
        this.afk = afk;
    }

    @Override
    public void setAFKMessage(String message) {

    }

    public void setAfkMessage(String message) {
        this.afkMessage = afkMessage;
    }

    public void setIgnore(String name, boolean ignore, boolean flagUpdate) {
        if (ignore) {
            this.ignores.add(name.toLowerCase());
        } else {
            this.ignores.remove(name.toLowerCase());
        }

        if (flagUpdate) {
            this.storage.flagUpdate(this);
        }
    }

    public void setLastPrivateMessageSource(Chatter chatter) {
        this.lastPMSource = chatter;
    }

    public void setMuted(boolean muted, boolean flagUpdate) {
        this.muted = muted;
        if (flagUpdate) {
            this.storage.flagUpdate(this);
        }
    }

    public boolean shouldAutoJoin(Channel channel) {
        return SurvivorBot.hasChannelPermission(this.player, channel, Permission.AUTOJOIN);
    }

    public boolean shouldForceJoin(Channel channel) {
        return SurvivorBot.hasChannelPermission(this.player, channel, Permission.FORCE_JOIN);
    }

    public boolean shouldForceLeave(Channel channel) {
        return SurvivorBot.hasChannelPermission(this.player, channel, Permission.FORCE_LEAVE);
    }

    public void refocus() {
        if (this.activeChannel != null) {
            if (this.canFocus(this.activeChannel) == Result.ALLOWED) {
                return;
            }

            this.lastActiveChannel = this.activeChannel;
            this.lastActiveChannel.onFocusLoss(this);
            this.activeChannel = null;
        }

        this.getChannels().stream().filter(channel -> canFocus(channel) == Result.ALLOWED).forEach(channel -> {
            this.activeChannel = channel;
            this.activeChannel.onFocusGain(this);

            try {
                Messaging.send(this.player, SurvivorBot.getMessage("chatter_focus"),
                        channel.getColor() + channel.getName());
            } catch (MessageNotFoundException ignored) {
                SurvivorBot.severe("Messages.properties is missing: chatter_focus");
            }
        });

    }

    public Result canIgnore(Chatter other) {
        return other.getPlayer().hasPermission("survivorbot.admin.unignore") ? Result.NO_PERMISSION : Result.ALLOWED;
    }

    public void disconnect() {
        this.channels.forEach(channel -> channel.removeMember(this, false, false));
    }
}

File: src/main/kotlin/com/fireblazerrr/survivorbot/chatter/StandardChatter.kt
package com.fireblazerrr.survivorbot.chatter

import com.fireblazerrr.survivorbot.SurvivorBot
import com.fireblazerrr.survivorbot.channel.Channel
import com.fireblazerrr.survivorbot.utils.message.Messaging
import net.md_5.bungee.api.ChatColor
import org.bukkit.Bukkit
import org.bukkit.entity.Player
import org.bukkit.scoreboard.Team


class StandardChatter(private val storage: ChatterStorage, override val player: Player) : Chatter {
    override var lastPMSource: Chatter = this
    override var activeChannel: Channel = SurvivorBot.channelManager.defaultChannel
    override var lastActiveChannel: Channel = SurvivorBot.channelManager.defaultChannel
    override var lastFocusableChannel: Channel = SurvivorBot.channelManager.defaultChannel
    override var afkMessage: String = ""
    override var muted: Boolean = false
    override var afk: Boolean = false
    override var team: Team? = Bukkit.getScoreboardManager().mainScoreboard.teams.firstOrNull { it.entries.contains(player.name) }

    private val channels: MutableSet<Channel> = HashSet()
    private val ignores: MutableSet<String> = HashSet()

    override fun addChannel(channel: Channel, announce: Boolean, flagUpdate: Boolean): Boolean {
        if (channels.add(channel)) {
            channel.addMember(this, announce, flagUpdate)

            if (flagUpdate) storage.flagUpdate(this)

            return true
        }
        return false
    }


    override fun canBan(channel: Channel): Chatter.Result = if (SurvivorBot.hasChannelPermission(player, channel, Chatter.Permission.BAN))
        Chatter.Result.ALLOWED
    else
        if (channel.isChatterModerator(player.name) && SurvivorBot.channelManager.checkModPermission(Chatter.Permission.BAN))
            Chatter.Result.ALLOWED
        else
            Chatter.Result.NO_PERMISSION


    override fun canColorMessage(channel: Channel, color: ChatColor): Chatter.Result = Chatter.Result.ALLOWED


    override fun canEmote(channel: Channel): Chatter.Result = if (!channel.isChatterMember(this.player.name))
        Chatter.Result.INVALID
    else
        if (!channel.isTransient() && SurvivorBot.hasChannelPermission(player, channel, Chatter.Permission.EMOTE))
            if (!muted && !channel.isChatterMuted(player.name))
                if (!channel.hasWorld(player.world))
                    Chatter.Result.BAD_WORLD
                else
                    Chatter.Result.ALLOWED
            else
                Chatter.Result.MUTED
        else
            Chatter.Result.NO_PERMISSION


    override fun canFocus(channel: Channel): Chatter.Result = if (!SurvivorBot.hasChannelPermission(player, channel, Chatter.Permission.FOCUS))
        Chatter.Result.NO_PERMISSION
    else {
        val speak: Chatter.Result = canSpeak(channel)

        if (speak != Chatter.Result.ALLOWED && speak != Chatter.Result.INVALID)
            Chatter.Result.NO_PERMISSION
        else
            Chatter.Result.ALLOWED
    }


    override fun canJoin(channel: Channel, password: String): Chatter.Result = if (channel.isChatterMember(player.name))
        Chatter.Result.INVALID
    else
        if (!SurvivorBot.hasChannelPermission(player, channel, Chatter.Permission.JOIN))
            Chatter.Result.NO_PERMISSION
        else
            if (channel.isChatterBanned(player.name))
                Chatter.Result.BANNED
            else
                if (password != channel.password)
                    Chatter.Result.BAD_PASSWORD
                else
                    Chatter.Result.ALLOWED


    override fun canKick(channel: Channel): Chatter.Result = if (SurvivorBot.hasChannelPermission(player, channel, Chatter.Permission.KICK))
        Chatter.Result.ALLOWED
    else
        if (channel.isChatterModerator(player.name) && SurvivorBot.channelManager.checkModPermission(Chatter.Permission.KICK))
            Chatter.Result.ALLOWED
        else
            Chatter.Result.NO_PERMISSION


    override fun canLeave(channel: Channel): Chatter.Result = if (!channel.isChatterMember(player.name))
        Chatter.Result.INVALID
    else
        if (SurvivorBot.hasChannelPermission(player, channel, Chatter.Permission.LEAVE))
            Chatter.Result.NO_PERMISSION
        else
            Chatter.Result.ALLOWED


    override fun canModify(setting: String, channel: Channel): Chatter.Result {
        val lowerSetting = setting.toLowerCase()
        val permission: Chatter.Permission = when (lowerSetting) {
            "nick" -> Chatter.Permission.MODIFY_NICK
            "format" -> Chatter.Permission.MODIFY_FORMAT
            "distance" -> Chatter.Permission.MODIFY_DISTANCE
            "color" -> Chatter.Permission.MODIFY_COLOR
            "shortcut" -> Chatter.Permission.MODIFY_SHORTCUT
            "password" -> Chatter.Permission.MODIFY_PASSWORD
            "verbose" -> Chatter.Permission.MODIFY_VERBOSE
            "chatcost" -> Chatter.Permission.MODIFY_CHATCOST
            "crossworld" -> Chatter.Permission.MODIFY_CROSSWORLD
            else -> {
                if (setting != "focusable")
                    return Chatter.Result.INVALID

                Chatter.Permission.MODIFY_FOCUSABLE
            }
        }

        return if (SurvivorBot.hasChannelPermission(player, channel, permission))
            Chatter.Result.ALLOWED
        else
            if (channel.isChatterModerator(player.name) && SurvivorBot.channelManager.checkModPermission(permission))
                Chatter.Result.ALLOWED
            else
                Chatter.Result.NO_PERMISSION
    }


    override fun canMute(channel: Channel): Chatter.Result = if (SurvivorBot.hasChannelPermission(player, channel, Chatter.Permission.MUTE))
        Chatter.Result.ALLOWED
    else
        if (channel.isChatterModerator(player.name) && SurvivorBot.channelManager.checkModPermission(Chatter.Permission.BAN))
            Chatter.Result.ALLOWED
        else
            Chatter.Result.NO_PERMISSION


    override fun canRemove(channel: Channel): Chatter.Result = if (SurvivorBot.hasChannelPermission(player, channel, Chatter.Permission.REMOVE))
        Chatter.Result.ALLOWED
    else
        if (channel.isChatterModerator(player.name) && SurvivorBot.channelManager.checkModPermission(Chatter.Permission.REMOVE))
            Chatter.Result.ALLOWED
        else
            Chatter.Result.NO_PERMISSION


    override fun canSpeak(channel: Channel): Chatter.Result = if (!channel.isChatterMember(player.name))
        Chatter.Result.INVALID
    else
        if (!channel.isTransient() && SurvivorBot.hasChannelPermission(player, channel, Chatter.Permission.SPEAK))
            Chatter.Result.NO_PERMISSION
        else
            if (!muted && !channel.isChatterMuted(player.name))
                if (!channel.hasWorld(player.world))
                    Chatter.Result.BAD_WORLD
                else
                    Chatter.Result.ALLOWED
            else
                Chatter.Result.MUTED


    override fun canViewInfo(channel: Channel): Chatter.Result = if (!SurvivorBot.hasChannelPermission(player, channel, Chatter.Permission.INFO))
        Chatter.Result.NO_PERMISSION
    else
        Chatter.Result.ALLOWED


    override fun canIgnore(other: Chatter): Chatter.Result = if (other.player.hasPermission("survivorbot.admin.unignore"))
        Chatter.Result.NO_PERMISSION
    else
        Chatter.Result.ALLOWED


    override fun isIgnoring(chatter: Chatter): Boolean = isIgnoring(chatter.player.name)
    override fun isIgnoring(chatter: String): Boolean = ignores.contains(chatter.toLowerCase())

    override fun isInRange(chatter: Chatter, range: Int): Boolean = (player.world == chatter.player.world && player.location.distanceSquared(chatter.player.location) <= (range * range))

    override fun removeChannel(channel: Channel, announce: Boolean, flagUpdate: Boolean): Boolean {
        if (channels.contains(channel)) {
            channels.remove(channel)
            if (channel.isChatterMember(player.name))
                channel.removeMember(this, announce, flagUpdate)

            if (flagUpdate)
                this.storage.flagUpdate(this)

            return true
        }
        return false
    }

    override fun setActiveChannel(channel: Channel, announce: Boolean, flagUpdate: Boolean) {
        if (channel != activeChannel) {
            activeChannel.onFocusLoss(this)

            if (!activeChannel.isTransient()) {
                lastActiveChannel = activeChannel
                if (canFocus(activeChannel) == (Chatter.Result.ALLOWED)) {
                    lastFocusableChannel = activeChannel
                }
            }

            activeChannel = channel
            activeChannel.onFocusGain(this)
            if (announce)
                Messaging.send(player, SurvivorBot.getMessage("chatter_focus"), "${channel.color}${channel.name}")
            if (flagUpdate)
                storage.flagUpdate(this)

        }
    }

    override fun setIgnore(name: String, ignore: Boolean, flagUpdate: Boolean) {
        if (ignore)
            ignores.add(name.toLowerCase())
        else
            ignores.remove(name.toLowerCase())


        if (flagUpdate)
            storage.flagUpdate(this)
    }

    override fun shouldAutoJoin(channel: Channel): Boolean = SurvivorBot.hasChannelPermission(player, channel, Chatter.Permission.AUTOJOIN)

    override fun shouldForceJoin(channel: Channel): Boolean = SurvivorBot.hasChannelPermission(player, channel, Chatter.Permission.FORCE_JOIN)

    override fun shouldForceLeave(channel: Channel): Boolean = SurvivorBot.hasChannelPermission(player, channel, Chatter.Permission.FORCE_LEAVE)

    override fun hasChannel(channel: Channel): Boolean = channels.contains(channel)

    override fun refocus() {
        if (canFocus(activeChannel) == Chatter.Result.ALLOWED)
            return
        lastActiveChannel = activeChannel
        lastActiveChannel.onFocusLoss(this)

        channels.filter { canFocus(it) == Chatter.Result.ALLOWED }.forEach{
            activeChannel = it
            activeChannel.onFocusGain(this)

            Messaging.send(player, SurvivorBot.getMessage("chatter_focus"), "${it.color}${it.name}")
        }
    }

    override fun disconnect() {
        val removal: Iterator<Channel> = channels.iterator()
        while (removal.hasNext()){
            val ch: Channel = removal.next()
            ch.removeMember(this, false, false)
        }
    }
}
